### PR TITLE
[BugFix] Fix bitmap function bugs when bitmap value may not be BITMAP type

### PR DESCRIPTION
**Pull request by LiShuMing**
Fixes #issue

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/types/bitmap_value.cpp [hunk 1]
```diff
@@ -922,92 +922,138 @@ void BitmapValue::_convert_to_smaller_type() {
     }
 }
 
-int64_t BitmapValue::sub_bitmap_internal(const int64_t& offset, const int64_t& len, BitmapValue* ret_bitmap) {
-    if (offset > 0 && offset >= _bitmap->cardinality()) {
-        return 0;
+void BitmapValue::_convert_to_bitmap() {
+    switch (_type) {
```
Comment (stdpain) [inline be/src/types/bitmap_value.cpp:926]: it was not expected

Comment (stdpain) [reply be/src/types/bitmap_value.cpp:926]: we shouldn't change input bitmap type

**Review by trueeyu — DISMISSED** [review]

File: be/src/types/bitmap_value.cpp (hunk 1 above)
Comment (trueeyu) [reply be/src/types/bitmap_value.cpp:926]: > we shouldn't change input bitmap type

If not convert to bitmap, maybe it's hard to get sub_bitmap by idx

**Comment by trueeyu** [comment]
Add one case on SQLTester?

File: be/src/types/bitmap_value.cpp (hunk 1 above)
Comment (trueeyu) [reply be/src/types/bitmap_value.cpp:926]: Max size of set is 32, no need to convert to bitmap.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/types/bitmap_value.cpp [hunk 2]
```diff
@@ -923,91 +923,198 @@ void BitmapValue::_convert_to_smaller_type() {
...
         auto it = _bitmap->begin();
         for (; it != _bitmap->end() && *it < range_start;) {
             ++it;
         }
-        for (; it != _bitmap->end() && count < abs_limit; ++it, ++count) {
+        int64_t count = 0;
+        for (; it != _bitmap->end() && *it < range_end; ++it, ++count) {
             ret_bitmap->add(*it);
         }
+        return count;
```
Comment (fzhedu) [inline be/src/types/bitmap_value.cpp:1116]: merge the same?

File: be/src/types/bitmap_value.cpp [hunk 3]
```diff
@@ -923,91 +923,198 @@ void BitmapValue::_convert_to_smaller_type() {
...
 int64_t BitmapValue::sub_bitmap_internal(const int64_t& offset, const int64_t& len, BitmapValue* ret_bitmap) {
-    if (offset > 0 && offset >= _bitmap->cardinality()) {
-        return 0;
-    }
-    if (offset < 0 && std::abs(offset) > _bitmap->cardinality()) {
+    switch (_type) {
+    case EMPTY:
         return 0;
+    case SINGLE: {
+        if (offset >= 1 || offset < -1 || len <= 0) {
```
Comment (fzhedu) [inline be/src/types/bitmap_value.cpp:930]: offset can only be 0 or -1 ?

File: be/src/types/bitmap_value.cpp [hunk 4]
```diff
@@ -923,91 +923,198 @@ void BitmapValue::_convert_to_smaller_type() {
...
-    if (offset < 0) {
-        abs_offset = _bitmap->cardinality() + offset;
-    }
+    case SET: {
+        if (offset > 0 && offset >= _set->size()) {
+            return 0;
+        }
+        if (offset < 0 && std::abs(offset) > _set->size()) {
+            return 0;
+        }
```
Comment (fzhedu) [inline be/src/types/bitmap_value.cpp:943]: `offset >= xx || abs(offset) > xx` ?

File: be/src/types/bitmap_value.cpp [hunk 5]
```diff
@@ -923,91 +923,198 @@ void BitmapValue::_convert_to_smaller_type() {
...
+        int64_t abs_offset = offset;
+        if (offset < 0) {
+            abs_offset = _set->size() + offset;
```
Comment (fzhedu) [inline be/src/types/bitmap_value.cpp:946]: why add, must abs_offset > 0?

File: be/test/exprs/bitmap_functions_test.cpp [hunk 6]
```diff
@@ -2588,6 +2698,116 @@ TEST_F(VecBitmapFunctionsTest, bitmap_subset_limit) {
     }
 }
 
+TEST_F(VecBitmapFunctionsTest, sub_bitmap_limit_special_cases) {
+    // empty bitmap
+    {
+        auto bitmap_column = BitmapColumn::create();
```
Comment (fzhedu) [inline be/test/exprs/bitmap_functions_test.cpp:2704]: create some cases where column is only null / const column

File: be/src/types/bitmap_value.cpp (hunk 2 above)
Comment (LiShuMing) [reply be/src/types/bitmap_value.cpp:1116]: not exact same for each type

File: be/src/types/bitmap_value.cpp (hunk 5 above)
Comment (LiShuMing) [reply be/src/types/bitmap_value.cpp:946]: yep. this is original codes. I just move it here.

File: be/src/types/bitmap_value.cpp (hunk 4 above)
Comment (LiShuMing) [reply be/src/types/bitmap_value.cpp:943]: fix it.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/types/bitmap_value.cpp [hunk 7]
```diff
@@ -923,91 +923,192 @@ void BitmapValue::_convert_to_smaller_type() {
...
+        if (_sv < range_start) {
+            return 0;
+        } else {
+            ret_bitmap->add(_sv);
+            return 1;
+        }
+    }
+    case SET: {
+        auto it = _set->begin();
+        for (; it != _set->end() && *it < range_start;) {
```
Comment (stdpain) [inline be/src/types/bitmap_value.cpp:1091]: we need sort for _set. _set is not stable

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/types/bitmap_value.cpp [hunk 8]
```diff
@@ -923,91 +923,197 @@ void BitmapValue::_convert_to_smaller_type() {
...
+        for (; it != set_values.end() && count < len; ++it, ++count) {
+            ret_bitmap->add(*it);
+        }
+        return count;
     }
-    for (; it != _bitmap->end() && count < len; ++it, ++count) {
-        ret_bitmap->add(*it);
+    default:
+        DCHECK_EQ(_type, BITMAP);
+        if ((offset > 0 && offset >= _bitmap->cardinality()) ||
```
Comment (stdpain) [inline be/src/types/bitmap_value.cpp:961]: _bitmap->cardinality is a heavy operator. we should consider about save it as a temp variable

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/types/bitmap_value.cpp (hunk 3 above)
Comment (trueeyu) [reply be/src/types/bitmap_value.cpp:930]: values=1  offset=-2, len=10 result is ?

File: be/src/types/bitmap_value.cpp [hunk 9]
```diff
@@ -923,91 +923,188 @@ void BitmapValue::_convert_to_smaller_type() {
...
 
-    int64_t count = 0;
-    int64_t offset_count = 0;
-    auto it = _bitmap->begin();
-    for (; it != _bitmap->end() && offset_count < abs_offset; ++it) {
-        ++offset_count;
+        int64_t count = 0;
+        int64_t offset_count = 0;
+        auto it = set_values.begin();
+        for (; it != set_values.end() && offset_count < abs_offset; ++it) {
```
Comment (trueeyu) [inline be/src/types/bitmap_value.cpp:953]: Use offset directly?

File: be/src/types/bitmap_value.cpp [hunk 10]
```diff
@@ -923,91 +923,188 @@ void BitmapValue::_convert_to_smaller_type() {
...
+    default:
+        DCHECK_EQ(_type, BITMAP);
+        int64_t count = 0;
+        if (limit < 0) {
+            int64_t abs_limit = -limit;
+            auto start = _bitmap->begin();
+            auto end = _bitmap->begin();
+
+            int64_t offset = 0;
+            for (; end != _bitmap->end() && offset < abs_limit && *end <= range_start;) {
```
Comment (trueeyu) [inline be/src/types/bitmap_value.cpp:1042]: /**
     * Returns the number of integers that are smaller or equal to x.
     */
    uint64_t rank(uint64_t x) const {

File: be/src/types/bitmap_value.cpp [hunk 11]
```diff
@@ -923,91 +923,188 @@ void BitmapValue::_convert_to_smaller_type() {
...
-        for (; end != _bitmap->end() && offset < abs_limit && *end <= range_start;) {
-            ++end;
-            ++offset;
+    switch (_type) {
+    case EMPTY:
+        return 0;
+    case SINGLE: {
+        if (_sv < range_start) {
+            return 0;
+        } else {
```
Comment (trueeyu) [inline be/src/types/bitmap_value.cpp:993]: if limit = 0?

File: be/src/types/bitmap_value.cpp [hunk 12]
```diff
@@ -923,91 +923,188 @@ void BitmapValue::_convert_to_smaller_type() {
...
-        if (offset == abs_limit) {
-            for (; end != _bitmap->end() && *end <= range_start;) {
-                ++start;
+    }
+    case SET: {
+        std::vector values(_set->begin(), _set->end());
+        std::sort(values.begin(), values.end());
+
+        int64_t count = 0;
+        if (limit < 0) {
```
Comment (trueeyu) [inline be/src/types/bitmap_value.cpp:1003]: Why limit < 0 ,

File: be/src/types/bitmap_value.cpp [hunk 13]
```diff
@@ -923,91 +923,188 @@ void BitmapValue::_convert_to_smaller_type() {
...
+int64_t BitmapValue::bitmap_subset_in_range_internal(const int64_t& range_start, const int64_t& range_end,
+                                                     BitmapValue* ret_bitmap) {
+    switch (_type) {
+    case EMPTY:
+        return 0;
+    case SINGLE: {
+        if (_sv < range_start) {
+            return 0;
+        } else {
+            ret_bitmap->add(_sv);
```
Comment (trueeyu) [inline be/src/types/bitmap_value.cpp:1079]: check range_end?

File: be/src/types/bitmap_value.cpp [hunk 14]
```diff
@@ -923,91 +923,188 @@ void BitmapValue::_convert_to_smaller_type() {
...
+            return 1;
+        }
+    }
+    case SET: {
+        std::vector values(_set->begin(), _set->end());
+        std::sort(values.begin(), values.end());
+        auto it = values.begin();
+        for (; it != values.end() && *it < range_start;) {
```
Comment (trueeyu) [inline be/src/types/bitmap_value.cpp:1087]: use index directly?

File: be/src/types/bitmap_value.cpp (hunk 10 above)
Comment (LiShuMing) [reply be/src/types/bitmap_value.cpp:1021]: To use this, `Roaring64MapSetBitForwardIterator` need support `move`  op, can I optimize this later?

File: be/src/types/bitmap_value.cpp (hunk 13 above)
Comment (LiShuMing) [reply be/src/types/bitmap_value.cpp:1058]: good

File: be/test/exprs/bitmap_functions_test.cpp (hunk 6 above)
Comment (LiShuMing) [reply be/test/exprs/bitmap_functions_test.cpp:2704]: it's been added in normal cases.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by stdpain — DISMISSED** [review]

File: be/src/types/bitmap_value.cpp [hunk 15]
```diff
@@ -923,91 +923,164 @@ void BitmapValue::_convert_to_smaller_type() {
...
-    int64_t abs_limit = limit;
-    if (limit < 0) {
-        is_reverse = true;
-        abs_limit = -limit;
+    switch (_type) {
+    case EMPTY:
+        return 0;
+    case SINGLE: {
+        if (_sv < range_start || limit == 0) {
+            return 0;
```
Comment (trueeyu) [inline be/src/types/bitmap_value.cpp:987]: if limit=-1 && _sv < range_start, the result is ?

File: be/src/types/bitmap_value.cpp [hunk 16]
```diff
@@ -923,91 +923,164 @@ void BitmapValue::_convert_to_smaller_type() {
...
+        std::vector values(_set->begin(), _set->end());
+        std::sort(values.begin(), values.end());
+
+        int64_t count = 0;
+        if (limit < 0) {
+            int64_t abs_limit = -limit;
+            auto it = std::lower_bound(values.begin(), values.end(), range_start);
+            for (; it != values.begin() && count < abs_limit; --it, ++count) {
+                ret_bitmap->add(*it);
+            }
```
Comment (fzhedu) [inline be/src/types/bitmap_value.cpp:1003]: end() is invalid, but begin() is valid, so I think there may be bugs

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by stdpain — DISMISSED** [review]

File: be/src/types/bitmap_value.cpp (hunk 16 above)
Comment (ZiheLiu) [reply be/src/types/bitmap_value.cpp:1003]: Is is possible that `it` is `values.end()`? If `range_start` is greater than all the values in `values`, then `it` will be `values.end()`.

File: be/src/types/bitmap_value.cpp [hunk 17]
```diff
@@ -923,91 +923,164 @@ void BitmapValue::_convert_to_smaller_type() {
...
-            ++end;
-            ++offset;
+    case SET: {
+        std::vector values(_set->begin(), _set->end());
+        std::sort(values.begin(), values.end());
+
+        int64_t count = 0;
+        if (limit < 0) {
+            int64_t abs_limit = -limit;
+            auto it = std::upper_bound(values.rbegin(), values.rend(), range_start);
```
Comment (ZiheLiu) [inline be/src/types/bitmap_value.cpp:1000]: Maybe `std::lower_bound(values.rbegin(), values.rend(), range_start, std::greater<>())`?

Comment (LiShuMing) [reply be/src/types/bitmap_value.cpp]: yep, you are right. I found the sql tester cannot cover this. I added more unit tests.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/types/bitmap_value.cpp [hunk 18]
```diff
@@ -923,91 +923,164 @@ void BitmapValue::_convert_to_smaller_type() {
...
-            ++end;
-            ++offset;
+    case SET: {
+        std::vector values(_set->begin(), _set->end());
+        std::sort(values.begin(), values.end());
+
+        int64_t count = 0;
+        if (limit < 0) {
+            int64_t abs_limit = -limit;
+            auto it = std::lower_bound(values.rbegin(), values.rend(), range_start, std::greater<uint64_t>());
```
Comment (trueeyu) [inline be/src/types/bitmap_value.cpp:1000]: If range_start< 0， the result is ?

File: be/src/types/bitmap_value.cpp [hunk 19]
```diff
@@ -923,91 +923,164 @@ void BitmapValue::_convert_to_smaller_type() {
...
+    default:
+        DCHECK_EQ(_type, BITMAP);
+        int64_t count = 0;
+        if (limit < 0) {
+            int64_t abs_limit = -limit;
+            auto start = _bitmap->begin();
+            auto end = _bitmap->begin();
+
+            int64_t offset = 0;
+            for (; end != _bitmap->end() && offset < abs_limit && *end <= range_start;) {
```
Comment (trueeyu) [inline be/src/types/bitmap_value.cpp:1021]: Why not use iter--;

**Review by trueeyu — APPROVED** [review]

File: be/src/types/bitmap_value.cpp (hunk 19 above)
Comment (LiShuMing) [reply be/src/types/bitmap_value.cpp:1021]: I'll optimize this later in the next pr.

**Review by ZiheLiu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**